vllm/triton_utils/jit_cache.py

```diff
@@ -0,0 +1,387 @@
+            self.device)["max_shared_mem"]
+        if self.kernel.metadata.shared > self.dev_max_shared:
+            raise OutOfResources(self.kernel.metadata.shared,
+                                 self.dev_max_shared, "shared memory")
+        self.module, self.function, self.n_regs, self.n_spills = (
+            driver.active.utils.load_binary(
+                self.kernel.name,
+                self.kernel.kernel,
+                self.kernel.metadata.shared,
+                self.device,
+            ))
+
+    def __call__(self, *args, **kwargs):
+        assert len(args) == 0
+
+        for arg_n, idx in self.update_args_index.items():
+            self.non_const_vals_lst[idx] = kwargs[arg_n]
+
+        if self.cache_launch_grid:
+            grid_0, grid_1, grid_2 = self.concrete_grid
+        else:
+            if self.grid_is_callable:
+                grid = kwargs["grid"](kwargs)
+            else:
+                grid = kwargs["grid"]
+            grid_size = len(grid)
+            grid_0 = grid[0]
+            grid_1 = grid[1] if grid_size > 1 else 1
+            grid_2 = grid[2] if grid_size > 2 else 1
+
+        stream = driver.active.get_current_stream(self.device)
+
+        return self.run(
+            grid_0,
+            grid_1,
+            grid_2,
+            stream,
+            self.function,
+            self.kernel.packed_metadata,
+            self.launch_metadata,
+            self.launch_enter_hook,
+            self.launch_exit_hook,
+            *self.non_const_vals_lst,
+        )
+
+    def get_key(self):
+        return self.cache_key
+
+
+class JitCache(KernelInterface):
+
+    def __init__(
+        self,
+        fn,
+        arg_names,
+        check_keys,
+        cache_lock: CacheLock,
+        cache_launch_grid=False,
+        assume_const=None,
+    ):
+        # we depend on the triton version, right now, 3.0 -- 3.2 are supported
+        assert (int(triton_version.split(".")[0]) == 3
+                and int(triton_version.split(".")[1]) <= 2)
+        self.arg_names = arg_names
+        self.fn = fn
+        self.base_fn = fn
+        while not inspect.isfunction(self.base_fn):
+            self.base_fn = self.base_fn.fn
+        self.cache_lock = cache_lock
+        self.cache_launch_grid = cache_launch_grid
+        self.dynamic_mode = False
+        self.run = self._run_static
+        if self.cache_lock is None:
+            self.dynamic_mode = True
+            self.run = self._run_dynamic
+        self.check_keys = check_keys
+        self.assume_const = assume_const
+        self.kernel_cache: dict[str, PreparedKernel] = {}
+
+        def calc_cache_index(kwargs):
+            cache_key = ""
+            for c_arg_name in check_keys:
+                cache_key += str(kwargs[c_arg_name])
+            return cache_key
+
+        self.cache_index_func = calc_cache_index
+        if len(check_keys) == 0:
+            self.cache_index_func = lambda ignore: "_default_"
```

Comment:
This seems like a scary default. Wouldn't this just result in lots of false hits to the cache? I.E all functions mapping to the first function added?

Reply:
I think the current name is fine but we should be very clear about our expectations of users of the jit cache. Like naively I would somewhat expect key generation to be "transparent" from the users, meaning we can just make a key given the values of all constexpr and assumed_const args. Though, maybe there is a performance concern with making a key with so many values?

Reply:
Yeah, if that's the case, then I think that `cache_keys` should be a non-optional argument.